### PR TITLE
feat(mktemp): add GNU-compatible mktemp command

### PR DESCRIPTION
**Pull request by trieloff**
## Summary

Adds `mktemp` so scripts can create a temporary file or directory and use the returned path immediately. For example, `f=$(mktemp) && process-data > "$f"` now creates an empty private file in a non-empty `$TMPDIR`, or `/tmp` otherwise.

The command supports `-d`, `-u`, `-q`, `-p DIR`, `--tmpdir[=DIR]`, `-t`, explicit and inferred suffixes, short-option clusters, `--help`, and `--version`. It preserves relative template output and follows GNU directory selection, including the different `$TMPDIR` precedence of `-t`.

## Exclusive creation

Adds optional `IFileSystem.createExclusive(path, { mode, directory })` and implements it in the in-memory, overlay, read-write, and mountable backends. Files are created with mode 0600 and directories with mode 0700. Names use Web Crypto with rejection sampling and bounded buffers.

The operation rejects occupied names, including dangling symlinks, and applies permissions during creation. The read-write backend uses exclusive file opens and `mkdir(mode)`. The defense-aware command context forwards the method, with interface invariant tests guarding against future omissions.

External filesystem implementations remain source compatible because the method is optional. `mktemp` fails clearly when a backend cannot provide exclusive creation; there is no check/write/chmod fallback. Only name collisions are retried.

The read-write backend revalidates the created entry's identity and containment. An intermediate-directory swap by a concurrent host process can still cause an empty entry to be created outside the root before validation reports failure. This detects such a race; it does not prevent it. Preventing it requires descriptor-relative filesystem operations beyond the Node APIs used here.

## Command coverage

- Empty `--tmpdir=` and `-p ''` fall back to `$TMPDIR` or `/tmp`.
- A separate directory argument works after `-dp`; missing arguments fail.
- `-q` still reports malformed templates.
- Earlier X runs remain literal, and `TMPDIR=/` prints a single leading slash.
- 12 GNU comparison fixtures check generated names' shape and location, file type, private mode, suffix handling, dry-run behavior, and failures. Random names are never recorded. Fixtures were recorded against GNU coreutils 9.7.

## Compatibility notes

- `-h` is accepted as an alias for `--help`; GNU rejects it.
- Unambiguous GNU long-option abbreviations are not implemented.
- `--version` currently prints `mktemp (just-bash) 9.4`.
- Dry-run mode probes candidate names with `lstat` but never creates entries.

## Validation

- 102 focused command, filesystem, and defense-context tests pass, including 11 new option/template tests.
- All 884 comparison tests pass, including the 12 new GNU cases.
- The executor package passes all 74 tests.
- Build, typecheck, lint, and knip pass.
- The broader just-bash suite reports 15,703 passed, 98 skipped, and 10 failed. All 10 failures reproduce in the same tests on an independently built, unmodified `main` at `062ce005`: six Python worker failures and four special-permission-bit failures on this macOS/Node 26 host.

Minor changeset included.

Original implementation generated with Claude Code; comparison coverage added with Codex.

### CODE REVIEW

**Comment by trieloff** [comment]
Thanks — the review was right on all the substantive points. Addressed in db407a9.

## The blocker: creation was neither exclusive nor private

All four reviewers converged on the same defect, and it was real. `exists()` → `writeFile()`/`mkdir()` → `chmod()` failed both of the guarantees the PR advertises:

- **TOCTOU.** The gap between probe and create let a concurrent creator win, and `writeFile` would then truncate its entry rather than retry.
- **Permission window.** The entry was born with the backend default (`ReadWriteFs.replaceFile` opens new files `0o666 & ~umask`; `mkdir` gives 0755) and only tightened afterwards — on a real-FS-backed root, long enough for another local process to open it and retain the fd. If the `chmod` itself threw, the loose-mode entry was left behind.
- **Symlinks.** The second-opinion reviewer's catch was the sharpest one: `exists()` resolves symlinks and reports a *dangling* link as absent, so a pre-planted link at a candidate name read as free and the subsequent write went through to its target. That contradicted "an existing path is never returned" outright.

Rather than patch the sequence, I added the primitive the reviews asked for:

```ts
createExclusive?(path: string, options: { mode: number; directory?: boolean }): Promise<void>
```

One operation that creates only if the name is free and applies the mode at creation time, never following the final path component:

- **ReadWriteFs** — `O_CREAT|O_EXCL|O_NOFOLLOW` with the mode passed to `open`, and `mkdir(mode)` for directories, so the kernel enforces exclusivity and the mode; POSIX makes `O_EXCL` fail on a symlink at the final component.
- **InMemoryFs / OverlayFs** — exclusive by construction (single-threaded, keyed by normalized path, so a symlink is a collision rather than a target). Overlay entries shadow the real FS underneath.
- **MountableFs** — delegates, and reports `ENOSYS` instead of silently degrading if a mounted implementation lacks the method.

It is optional on the interface, matching the existing `readFileBytes?` convention, so external `IFileSystem` implementations keep working; they get a documented best-effort fallback that at least probes with `lstat` (not `exists`) and removes the entry if `chmod` fails. `mktemp` no longer chmods at all on the built-in backends.

## The two GNU-compatibility bugs

Both confirmed and fixed:

- `--dry-run` no longer runs the parent-directory check, so `mktemp -u -p /missing fooXXX` prints a candidate. I kept the existence probe for `-u` because gnulib's `GT_NOCREATE` path does the same thing — it `lstat`s and retries — it just never creates.
- `--` now terminates option parsing, so `mktemp -- --help` and `mktemp -- --version` treat the argument as a template.

## Tests

+18 tests (15402 passing). The security guarantees are pinned at the filesystem layer, where they can be asserted deterministically rather than hoping a random 62³ name space collides:

- `read-write-fs.create-exclusive.test.ts` — mode 0600/0700 present at creation on the real FS; EEXIST on an occupied name with contents intact; a symlink (and a dangling symlink) at the name rejected **with the target verified untouched**; ENOENT on a missing parent.
- `interface.contract.test.ts` — the same contract for the in-memory backend, including the symlink-collision case.
- `mktemp.test.ts` — the `-u`/`--` regressions above, plus an end-to-end uniqueness check that runs 50 `mktemp` calls, writes a distinct marker through each returned path and asserts 50 distinct files with 50 distinct markers.

Two notes for whoever picks this up:

- I did **not** add a mktemp-level test that pre-plants a symlink at a candidate name — with a minimum of three X's the smallest name space is 62³, so such a test would be probabilistic theatre. The guarantee is tested at the layer that actually enforces it.
- `pnpm test:run` reports 6 failures, all CPython-WASM tests timing out at the 5s/30s limits on my machine; they reproduce identically on unmodified `main`.

**Comment by trieloff** [comment]
Addressed in bdf0108. The review ran against `db407a9`, so it predates 212892c, which forwarded `createExclusive` through the defense-aware command context.

Seven findings fixed, one consciously accepted.

## Option parsing

Help and version detection ran over raw argv, so an option *value* could trigger it. Both `mktemp -p --help` and `mktemp --suffix --version fileXXX` now consume the argument as a value. The scan that already skipped values for `--tmpdir` is shared for this rather than duplicated.

`getRandomValues` rejects buffers over 65536 bytes, so a template with a longer X run threw outside the command's error handling. The random part is now drawn in bounded chunks; a 70000-X template is covered by a test.

## InMemoryFs

`createExclusive` resolved nothing, so with a symlinked temp directory the entry was stored under the unresolved key while every later lookup resolved through the link — mktemp reported success and returned a path that did not exist. It now resolves symlinks in the parent while leaving the final component literal, which keeps a symlink at the name itself a collision rather than a target.

## OverlayFs

Both findings were correct.

The interleaving is real despite the single thread: the `await` on `existsInOverlay` yields, so concurrent calls could all observe absence. The check and the insert are now separated by no await. The regression test asserts exactly one of three concurrent calls succeeds, and it fails without the guard — I verified that by reverting the guard rather than assuming.

The tombstone was simply missing. Recreating a path that had been removed from the real layer succeeded while remaining invisible to `stat`, `exists` and `readdir`. `this.deleted.delete(normalized)` now runs, as it does on the adjacent `mkdir` and `writeFile` paths.

## ReadWriteFs parent swap

Correct, and the sharpest finding. `O_EXCL` and `O_NOFOLLOW` bind only the final component, so a concurrent host process swapping an intermediate directory for a symlink could place the entry outside the root.

Creation is now re-validated afterwards, comparing device and inode against the handle that was created, and a failure removes the entry before the error surfaces. To be precise about the guarantee: this **detects and reverts** an escape rather than preventing it, since Node exposes no `openat` for a descriptor-anchored walk. The window is now bounded by the create itself rather than lasting until the write, and it matches the re-validate-after-mkdir idiom already used in `writeFile`. If you would rather have prevention, that needs a dirfd-based primitive and is a larger change than this PR should carry.

## MountableFs ENOSYS

Correct, and it reintroduced exactly the unclean failure this PR set out to remove. `createExclusiveOn` now treats ENOSYS as "not supported" and degrades to the fallback, so mktemp works on a mount whose backend predates the method. Covered by a test using a `MountableFs` over a backend without `createExclusive`.

## Consciously accepted: the fallback is not exclusive

Flagged as high, and it is accurate: `lstat` followed by `writeFile` can be raced, and the entry is briefly at the backend default mode.

I do not think it can be fixed, only removed. Atomicity cannot be synthesised from non-atomic primitives, so for an external `IFileSystem` that does not implement `createExclusive` the options are a best-effort create or no `mktemp` at all. The fallback runs only on filesystems that lack the method — every built-in backend takes the exclusive path — and it is documented as best effort at the one place it lives. Happy to make it throw instead if you would rather external implementations be forced to implement the method.

## Validation

- [x] 39 command tests, 16 filesystem tests, plus the defense-context invariants
- [x] `pnpm test:run` passes 15415 tests
- [x] typecheck, lint, knip pass
- [x] the OverlayFs concurrency guard verified load-bearing by reverting it

Six CPython-WASM tests fail on my machine at the 5s and 30s timeouts and reproduce identically on unmodified `main`.

**Comment by trieloff** [comment]
## Review status

Running record of the automated review findings and their disposition, kept in
this one comment and updated in place as further rounds land. Current head is
`c2d9492`, rebased on `main` at `fix(fs): run lazy file providers in the defense-in-depth trusted scope (#397)`. Everything below is fixed unless marked otherwise.

## Exclusive creation

The original implementation created temporary entries with `exists()` then
`writeFile()`/`mkdir()` then `chmod()`, which failed both guarantees `mktemp`
exists to provide. The check-then-create gap is a TOCTOU window where a
concurrent creator's entry gets truncated rather than retried; the
create-then-chmod gap publishes the entry at the backend default, which on a
real-filesystem root is long enough for another process to open it and keep the
descriptor; and `exists()` resolves symlinks, so a link planted at a candidate
name read as free and the write followed it to its target.

`IFileSystem.createExclusive(path, { mode, directory })` replaces that sequence
with one operation that creates only when the name is free, applies the mode at
creation time, and never follows the final component.

## ReadWriteFs

- creates with `O_CREAT|O_EXCL|O_NOFOLLOW` and `mkdir(mode)`, so the kernel
  enforces exclusivity and the mode
- re-validates after creating, comparing device and inode against a descriptor
  held on the entry it made, because `O_EXCL` and `O_NOFOLLOW` bind only the
  final component and a concurrent process can swap an intermediate directory
  for a symlink
- identifies a new directory through a descriptor opened `O_DIRECTORY` and
  `O_NOFOLLOW`, rather than trusting the pathname, so a replacement at the same
  path is caught instead of accepted
- performs no cleanup when that validation fails. An earlier revision removed
  the entry with a recursive `rm`, which was worse than the problem: that
  branch is reached only once the pathname has been proven untrustworthy, so
  the delete could follow the swapped component out of the root. The stray
  entry is empty, 0600, and at a path the attacker already controls; Node
  exposes no unlink by descriptor, so there is no safe removal to perform

To be precise about the guarantee: this detects and reports an escape rather
than preventing one. Prevention needs a descriptor-anchored walk that Node does
not expose, and would be a larger change than this PR should carry.

## OverlayFs and InMemoryFs

- both resolve symlinks in the parent while leaving the final component
  literal. Storing under an unresolved key produced an entry that later
  lookups, which do resolve, could not find — with a symlinked `TMPDIR`,
  `mktemp` reported success and returned a path that did not exist
- both require the parent to be a directory, so a child beneath a regular file
  returns ENOTDIR
- OverlayFs separates its existence check and its insert by no `await`. The
  awaits above yield, so concurrent calls could all observe absence and all
  claim the name. The regression test asserts exactly one of three concurrent
  calls succeeds, and fails without the guard
- OverlayFs clears the deletion tombstone, as the adjacent `mkdir` and
  `writeFile` paths do. Recreating a path removed from the real layer otherwise
  succeeded while staying invisible to `stat`, `exists` and `readdir`

`OverlayFs` blocks symlinks by default, so its symlinked-parent case arises
only for embedders passing `allowSymlinks: true`.

## Error dispatch

Backend diagnostics embed the caller-supplied path, so matching message text
let a crafted template impersonate an errno:

- `MountableFs` throws a typed `ExclusiveCreateUnsupportedError` and
  `createExclusiveOn` dispatches on the type. Substring-matching `ENOSYS` meant
  `mktemp /tmp/ENOSYS-XXXXXX` could turn a genuine `EEXIST` into a silent
  downgrade to the weaker create, under exactly the racing conditions that
  justify the atomic path
- `mktemp` matches errno by `error.code` or an anchored `EEXIST:` prefix, so
  unrelated failures on such a path are no longer retried 100 times and
  reported as "File exists"

`MountableFs` always defines `createExclusive` but reports ENOSYS when the
backend it routes to lacks it; `createExclusiveOn` treats that as unsupported
and degrades, rather than hard-failing on mounted external filesystems.

## Defense-in-depth

The defense-aware command context rebuilds the filesystem from an explicit
allowlist and re-adds optional members only when they are listed.
`createExclusive` was not, so with defense-in-depth enabled — the default —
commands saw the method as absent and took the fallback. The hardening was
inert in the configuration meant to be hardest.

The method is forwarded now, and the class of bug is guarded rather than the
instance: a `Record<keyof IFileSystem, true>` in the defense-context invariant
tests fails typecheck when a method is added to the interface without being
listed, and the test then asserts the wrapper forwards it.

## Option parsing and GNU compatibility

- help and version are detected in parser order, over options only. Operands
  are not scanned, so a template is never read as a short cluster — an earlier
  revision printed help for `mktemp chartXXXX`, because its second character
  is `h`. Scanning continues past operands, so `mktemp fooXXXX --help` reaches
  the option as GNU permutes; `mktemp -- --help` treats it as a template,
  `mktemp -p --help` and `mktemp -ph` as the directory, and
  `mktemp --bad --help` reports the invalid option
- `--help=x` and `--version=x` are rejected, since GNU's no-argument options
  allow no attached value
- only ENOENT means a candidate name is free. An EACCES or ELOOP from `lstat`
  says the name could not be inspected, and reporting an unverified path as
  available is the same class of failure as returning an occupied one
- `--dry-run` performs no filesystem access at all, so `mktemp -u -p /missing
  fooXXX` prints a candidate. The existence probe it does keep matches gnulib's
  `GT_NOCREATE` path, which lstats and retries without creating
- `-t` prefers a non-empty `$TMPDIR` over `-p`, unlike every other branch,
  which prefers `-p`
- long runs of `X` are drawn from `getRandomValues` in bounded chunks. Web
  Crypto rejects buffers over 65536 bytes, so a longer run threw outside the
  command's error handling

## Filesystems without the primitive

There is no fallback. Successive rounds kept finding the same shape of defect
in one — a concurrent actor taking the name and having its entry truncated, a
symlink installed so the write landed on its target, a trailing `chmod`
applying to a replacement — because exclusivity and privacy cannot be
synthesised from separate check, create and chmod calls. A caller that asked
for an exclusive private entry and silently received a racy one is worse off
than one told the filesystem cannot provide it, so `createExclusiveOn` throws
`ExclusiveCreateUnsupportedError`.

Every built-in filesystem implements the primitive and `ProcessSubstitutionFs`
inherits it from `InMemoryFs`, so this only affects external `IFileSystem`
implementations, which get a clear diagnostic and a non-zero exit rather than a
weaker guarantee than the one advertised.

`MountableFs` materializes a parent that exists only because a child is mounted
under it: `stat` reports such a directory and `writeFile` creates through it, so
`mktemp -p /mnt` should not fail on a path `stat` calls a directory.

## Not done

GNU accepts unambiguous long-option abbreviations such as `--tmpd=/tmp` and
`--direc`. This PR does not, because option parsing goes through the shared
`parseArgs`, which matches long names exactly for every command — `ls --rec`
is rejected today too. Teaching only `mktemp` to permute abbreviations would
make it the odd one out; teaching `parseArgs` would change every command at
once, which is a larger change than this PR should carry. Happy to do that
separately if it is wanted.

## Validation

- [x] 51 command tests and 18 filesystem tests
- [x] `pnpm test:run` passes 15684 tests
- [x] typecheck, lint, knip pass
- [x] the OverlayFs concurrency guard and the interface-parity guard both
      verified load-bearing by reverting them

Six CPython-WASM tests fail at the 5s and 30s timeouts on my machine and
reproduce identically on unmodified `main`.

File: packages/just-bash/src/commands/mktemp/mktemp.ts [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    if (stopped || result || !isOption) return;
+
+    if (arg.startsWith("--")) {
+      const name = arg.slice(2).split("=")[0];
+      if (name === "help") result = "help";
+      else if (name === "version") result = "version";
+      else if (!longs.has(name)) stopped = true;
+      return;
+    }
+
```
[the remark this answers is not in the document]
Comment (trieloff) [reply packages/just-bash/src/commands/mktemp/mktemp.ts:168]: Good catch — confirmed and fixed in e50987df.

`mktemp -u -ph` printed help instead of creating in the directory `h`, so the scan disagreed with the parser, which takes the rest of the token as `-p`'s value. The scan now stops at a value-taking short option, since everything after it in the token is that value. A boolean cluster still short-circuits, so `mktemp -dh` prints help.

Tests cover `-ph`, `-dp/work` and `-dh`.

**Comment by mutewinter** [comment]
Downstream note from an embedder: we ship our own `mktemp` today because the model reaches for it by reflex and finds nothing, and this PR retires that entirely, since `TMPDIR` pointed at a directory inside the mount layout is all our host needs. I closed my #392 in favor of this one; it had the same command surface without the exclusive-create plumbing the review rightly asked both of us for, and yours has it across all four backends. I read through `create-exclusive.ts` and the read-write backend's revalidation and have nothing to add.

**Comment by trieloff** [comment]
Same deal on our end, @mutewinter. I gave up and shipped my own overlay (with a trip wire, so CI fails on the renovate PR when a built-in lands).
